File: windows/PabloCompanion.Tests/Services/UrlValidatorTests.cs
using PabloCompanion.Services;

namespace PabloCompanion.Tests.Services;

public class UrlValidatorTests
{
    [Fact]
    public void ValidateScheme_HttpsUrl_ReturnsNull()
    {
        Assert.Null(UrlValidator.ValidateScheme("https://api.pablo.health"));
    }

    [Fact]
    public void ValidateScheme_HttpUrl_ReturnsError()
    {
        // In non-debug builds, plain HTTP should fail (unless localhost)
        var result = UrlValidator.ValidateScheme("http://evil.com");
        // In debug mode, this will also fail (non-localhost)
        Assert.NotNull(result);
    }

    [Fact]
    public void ThrowIfInvalid_HttpUrl_Throws()
    {
        Assert.Throws<UriFormatException>(() =>
            UrlValidator.ThrowIfInvalid("http://evil.com"));
    }

    [Theory]
    [InlineData("https://example.com/dashboard", "https://example.com")]
    [InlineData("https://example.com/dashboard/settings", "https://example.com")]
    [InlineData("https://example.com", "https://example.com")]
    [InlineData("https://example.com/", "https://example.com")]
    [InlineData("https://example.com:8080/path", "https://example.com:8080")]
    public void NormalizeToOrigin_StripsPath(string input, string expected)
    {
        Assert.Equal(expected, UrlValidator.NormalizeToOrigin(input));
    }

    [Fact]
    public void NormalizeToOrigin_InvalidUrl_ReturnsTrimmed()
    {
        Assert.Equal("not-a-url", UrlValidator.NormalizeToOrigin("not-a-url/"));
    }
}

File: mac/PabloCompanionTests/URLValidatorTests.swift
import Foundation
@testable import Pablo
import Testing

@Suite("URLValidator")
struct URLValidatorTests {
    @Test func httpsURLIsValid() {
        #expect(URLValidator.validateScheme("https://api.pablo.health") == nil)
    }

    @Test func httpsWithPathIsValid() {
        #expect(URLValidator.validateScheme("https://api.pablo.health/api/v1") == nil)
    }

    @Test func httpURLIsRejectedInRelease() {
        // In DEBUG builds, http://localhost is allowed; all other http is rejected.
        let result = URLValidator.validateScheme("http://evil.example.com")
        #expect(result != nil)
    }

    @Test func emptyURLIsRejected() throws {
        let result = URLValidator.validateScheme("")
        #expect(result != nil)
        #expect(try #require(result?.contains("empty")))
    }

    @Test func invalidURLFormatIsRejected() {
        let result = URLValidator.validateScheme("not a url at all :::")
        #expect(result != nil)
    }

    @Test func throwIfInvalidThrowsOnBadScheme() {
        #expect(throws: URLValidationError.self) {
            try URLValidator.throwIfInvalid("http://evil.example.com")
        }
    }

    @Test func throwIfInvalidPassesForHTTPS() throws {
        try URLValidator.throwIfInvalid("https://api.pablo.health")
    }
}
